drivers: adc: Add ADS1119 driver

kurddt opened the pull request:
The driver support both synchronous and asynchronous reading.

Only one channel is supported but the single or differential input can be configured

Only one-shot mode is used even when extra samples need to be read.

The PR also add `ADC_ACQ_TIME_MILLISECONDS` unit. It uses the '0' value which was not used, this does implies that checking of the unit was set or not is not possible anymore. I don't think this is an issue since the only in-tree driver checking the provided unit is https://github.com/zephyrproject-rtos/zephyr/blob/7ab16c457b304ccff82056243e1cee8913263d3e/drivers/adc/adc_lmp90xxx.c#L361 on which this driver is based on

Commit:
drivers: adc: Add ADS1119 driver

The driver support both synchronous and asynchronous reading.

Signed-off-by: Guillaume Lager <g.lager@innoseis.com>

kurddt commented:
Anyone available to review this one ?

XenuIsWatching reviewed (changes requested):
I don't have authority to merge in, but I took some time to review your code.

XenuIsWatching commented on drivers/adc/adc_ads1119.c:
```diff
@@ -0,0 +1,522 @@
+		return -EINVAL;
+	}
+
+	switch (acq_value) {
```
looking more into this, microseconds won't fit into the 14bits used by acq_time with the longest time. (Value range is 0-16383 for a given unit)

You have to work around this unfortunately... just provide a comment explaining it...

kurddt replied:
I introduced ADC_ACQ_TIME_MILLISECONDS

kurddt replied:
done

kurddt commented on drivers/adc/adc_ads1119.c:
```diff
@@ -0,0 +1,522 @@
+}
+#endif
+
+_Noreturn static void ads1119_acquisition_thread(struct device *dev)
```
The _Noreturn is part of C11 but it's not used anywhere else in zephyr so I will remove it

kurddt commented:
@XenuIsWatching Thanks for your review

kurddt commented on drivers/adc/adc_ads1119.c:
```diff
@@ -0,0 +1,520 @@
+		.i2c_dt_device_addr = DT_INST_REG_ADDR(n),	     \
+	};							     \
+								     \
+	static struct ads1119_data data_##n = {			     \
```
Found out that initialization use 1024B of ROM probably initializing the context in `ads1119_init` will use less ressources

kurddt replied:
It's because the stack is allocated inside the ads1119_data structure. moving it outside should reduce the amount of ROM needed to initialize the structure

Commit:
drivers: adc: Add runtime adc context initialization function

In some case the structure enclosing the adc context
can be quite big, making static initialization quite
resource consuming

Signed-off-by: Guillaume Lager <g.lager@innoseis.com>

ycsin commented on drivers/adc/adc_ads1119.c:
```diff
@@ -0,0 +1,528 @@
+	uint8_t config = 0;
+	int dr = 0;
+
+	if (channel_cfg->channel_id != 0) {
```
"_Only one channel is supported..._"

Do you mean only the channel 0 is supported?

kurddt replied:
Yes, but the input of the channel is configurable

ycsin commented on drivers/adc/adc_ads1119.c:
```diff
@@ -0,0 +1,528 @@
+static const struct adc_driver_api api = {
+	.channel_setup = ads1119_channel_setup,
+	.read = ads1119_read,
+	.ref_internal = 2048,
```
I think have this 2048 declared somewhere in a macro looks better

kurddt replied:
Done

XenuIsWatching commented:
I have submitted a pull request for a similar (but different) part for the family of ti i2c adc which are the ads1x1x. This includes the ads1013, ads1014, ads1015, ads1113, ads1114, and ads1115. Looking at the data sheet of the ads1119, it appears it is just so different from this family so it's best it says with it's own separate c file as is.

https://github.com/zephyrproject-rtos/zephyr/pull/40477

Some differences I've noticed between our two drivers are below
* build all adc tests
* ticks used for resolutions rather than; my comment here - (https://github.com/zephyrproject-rtos/zephyr/pull/37397#discussion_r724635092)
* thread used for both async and sync compiles

kurddt commented:
@anangl Could you maybe advice on the introduction of `ADC_ACQ_TIME_MILLISECONDS` versus the use of `ADC_ACQ_TIME_TICKS`

kurddt commented:
Can the stale label be removed ?

mbolivar-nordic commented:
@anangl please review when convenient

XenuIsWatching commented:
I don't know where @anangl is as this PR is now on the last page of the PR... 😢 

I did see someone ask if my ads1x1x driver is compatible with the ads1119... which it's not... but that means people are still interested in this part...
https://github.com/zephyrproject-rtos/zephyr/discussions/44920

For review comments, I'd suggest you take a look at the ads1x1x driver... as I did get a lot of comments about issues that are now fixed in what got upstreamed that I currently still see in the ads1119 driver

XenuIsWatching commented on drivers/adc/adc_ads1119.c:
```diff
@@ -0,0 +1,519 @@
+
+	adc_context_start_read(&data->ctx, sequence);
+
+	if (wait) {
```
I'm still not convinced `wait` is needed here, after experimenting myself

https://github.com/zephyrproject-rtos/zephyr/blob/64af112d30cd7340f2a32a4eaebe86cd995bdd06/drivers/adc/adc_ads1x1x.c#L511

kurddt replied:
The reason why this is working in your driver is because the additional thread is created even if CONFIG_ADC_ASYNC` is set to `n` which is not the case in this one. When a single thread is used we need to make that the sync semaphore is only taken if already given

XenuIsWatching commented on drivers/adc/adc_ads1119.c:
```diff
@@ -0,0 +1,522 @@
+	const struct ads1119_config *config = dev->config;
+	struct ads1119_data *data = dev->data;
+
+	adc_context_init(&data->ctx);
```
not necessary, remove

kurddt replied:
I do think it's necessary otherwise the adc context will not be initialized

XenuIsWatching commented on drivers/adc/adc_ads1119.c:
```diff
@@ -0,0 +1,522 @@
+		return -ENODEV;
+	}
+
+	rc = ads1119_read_reg(dev, ADS1119_REG_STATUS, &status);
```
is this used anywhere? If not, remove it

kurddt replied:
It is used to check that the adc can be accessed

XenuIsWatching commented on drivers/adc/adc_ads1119.c:
```diff
@@ -0,0 +1,522 @@
+			    (.stack = thread_stack_##n),					 \
+			    ())									 \
+	};											 \
+	static struct ads1119_data data_##n;							 \
```
the timer, lock, and sync should be initialized here, remove `adc_context_init` function
https://github.com/zephyrproject-rtos/zephyr/blob/64af112d30cd7340f2a32a4eaebe86cd995bdd06/drivers/adc/adc_ads1x1x.c#L620

kurddt replied:
Doing so would mean that the whole ads1119_data structure will need to be allocated in the data section, which might be a waste of ROM. 
This is especially true when the whole structure is much bigger that the element that need to be initialized with something else than 0.
I think the ads1x1x driver would greatly benefit to not be statically initialized since the stack of the thread is a member of the structure. Meaning than a part of the ROM is only used to store the zeros to initialize the whole stack

XenuIsWatching replied:
While I understand your point...that could be a discussion for a separate PR. This is the way _all_ ADCs handle it for now. not just the ads1x1x. It should be done within the data for now.
Perhaps @anangl can comment on why it's done this way.

kurddt replied:
I will wait @anangl to comment on that

anangl replied:
There is no requirement to initialize the timer, lock, and async fields statically. I agree with @kurddt. It makes sense to initialize them at runtime when the data structure is big and all other fields can use the default initialization to 0.

XenuIsWatching commented on include/zephyr/dt-bindings/adc/adc.h:
```diff
@@ -14,6 +14,8 @@
 #endif
 
+/** Acquisition time is expressed in milliseconds. */
+#define ADC_ACQ_TIME_MILLISECONDS  (0)
```
remove and use ticks for now

XenuIsWatching commented:
You may be able to add support for the ads1219 with your driver with some macrobatics 🧙 as the only difference between the ads1119 and ads1219 is the resolution while the reg module is the same

henrikbrixandersen commented on drivers/adc/Kconfig.ads1119:
```diff
@@ -0,0 +1,35 @@
+
+if ADC_ADS1119
+
+config ADC_ADS1119_INIT_PRIORITY
+	int "ADC ADS1119 driver initialization priority"
+	default 70
```
FWIW, the `adc_lmp90xxx.c` ADC driver is for a SPI ADC. It uses the common `CONFIG_ADC_INIT_PRIORITY` without any issues so far.

kurddt replied:
This might be problematic when two different ADC parts are used on the same board and requires different init priority but will do

kurddt replied:
I seems that the ads1x1x also uses a dedicated init priority config

kurddt replied:
Done

XenuIsWatching commented on drivers/adc/adc_ads1119.c:
```diff
@@ -0,0 +1,522 @@
+
+#define ADS1119_REG_SHIFT 2
+
+#define ADS1119_RESOLUTION 15
```
See discussion here about the 16 bit resolution definition for differential and single-ended
https://github.com/zephyrproject-rtos/zephyr/pull/40477#discussion_r765875293

How I had to implement this was to cache if it was configured for a differential and then validate the configured resolution.
https://github.com/zephyrproject-rtos/zephyr/blob/22fd253fead279aa3c33a5dda51ada410a11782d/drivers/adc/adc_ads1x1x.c#L450

kurddt commented on drivers/adc/adc_ads1119.c:
```diff
@@ -0,0 +1,522 @@
+	uint16_t acq_value = ADC_ACQ_TIME_VALUE(acq_time);
+
+	if (acq_time != ADC_ACQ_TIME_DEFAULT &&
+	    ADC_ACQ_TIME_UNIT(acq_time) != ADC_ACQ_TIME_MILLISECONDS) {
```
Using TICKS unit means that the application do need to know exactly what ADC part is used, which for me defeat the purpose of the ADC generic API.

XenuIsWatching replied:
Yes, I agree... but I wasn't happy with using ticks either. but you wrote the define as `0` which is probably intended for undefined value..
It probably should use TICKS for now... and more 'grander' discussion could be started in an Issue about how to make the ADC generic API more generic, as now 3 (including this one) have acq times that don't fit into the generic API.

kurddt commented on drivers/adc/adc_ads1119.c:
```diff
@@ -0,0 +1,522 @@
+	.read = ads1119_read,
+	.ref_internal = ADS1119_REF_INTERNAL,
+#ifdef CONFIG_ADC_ASYNC
+	.read_async = ads1119_adc_read_async,
```
Fixed

kurddt commented on drivers/adc/adc_ads1119.c:
```diff
@@ -0,0 +1,522 @@
+	}
+
+	if ((status & ADS1119_STATUS_MASK_READY)
+	    >> ADS1119_STATUS_SHIFT_READY == 1) {
```
Fixed

kurddt commented:
> You may be able to add support for the ads1219 with your driver with some macrobatics mage as the only difference between the ads1119 and ads1219 is the resolution while the reg module is the same

I would rather have this merged first

anangl commented on tests/drivers/build_all/adc/testcase.yaml:
```diff
@@ -65,3 +65,7 @@ tests:
   drivers.adc.test.build:
     platform_allow: qemu_cortex_m3
     tags: adc_test
+  drivers.adc.i2c.build:
```
There is already such test case defined for ADS1X1X, so I think these two should be merged into one.

anangl commented on tests/drivers/build_all/adc/app.overlay:
```diff
@@ -186,5 +186,22 @@
 			};
 		};
+
+		test_i2c: i2c@11112222 {
```
Such node already exists, so `test_i2c_ads1119` can be added there.

anangl commented on drivers/adc/adc_ads1119.c:
```diff
@@ -0,0 +1,499 @@
+		    ())										 \
+	static const struct ads1119_config config_##n = {					 \
+		.bus = I2C_DT_SPEC_GET(DT_DRV_INST(n)),						 \
+		COND_CODE_1(CONFIG_ADC_ASYNC,							 \
+			    (.stack = thread_stack_##n),					 \
+			    ())									 \
```
```suggestion
		IF_ENABLED(CONFIG_ADC_ASYNC, (.stack = thread_stack_##n,))			 \
```

XenuIsWatching commented on drivers/adc/adc_ads1119.c:
```diff
@@ -0,0 +1,499 @@
+struct ads1119_config {
+	const struct i2c_dt_spec bus;
+	k_thread_stack_t *stack;
+	uint8_t resolution;
```
is this used anywhere?

kurddt commented on drivers/adc/adc_context.h:
```diff
@@ -110,6 +110,15 @@ static void adc_context_on_timer_expired(struct k_timer *timer_id)
 #endif /* ADC_CONTEXT_USES_KERNEL_TIMER */
 
 
+static inline void adc_context_init(struct adc_context *ctx)
```
That would require to also move `adc_context_on_timer_expired`. Either splitting the timer related functions in two blocks or moving them above `adc_context_request_next_sampling`. Let me know what you prefer

anangl replied:
You can just declare `adc_context_on_timer_expired` before `adc_context_init`.

kurddt commented on drivers/adc/adc_ads1119.c:
```diff
@@ -0,0 +1,499 @@
+	rc = ads1119_adc_start_read(dev, sequence, false);
+
+
+	while (rc == 0 && k_sem_count_get(&data->ctx.sync) == 0) {
+		rc = ads1119_adc_perform_read(dev);
+	}
+	k_sem_take(&data->ctx.sync, K_FOREVER);
```
But I might change the `k_sem_count_get` and `k_sem_take` outside the loop with you suggestion:
```
while (rc == 0 && k_sem_take(&data->ctx.sync, K_NO_WAIT) != 0) {
		rc = ads1119_adc_perform_read(dev);
}
```

anangl replied:
> But I do think that it's harder to understand

Both ways are quite unclear, actually. You can always add some comment to make it easier to understand. I tried to make it consistent with what happens in `adc_context_wait_for_completion`. But I won't insist on changing this. It's up to you how you want it to be.

XenuIsWatching commented on drivers/adc/adc_context.h:
```diff
@@ -110,6 +122,8 @@ static void adc_context_on_timer_expired(struct k_timer *timer_id)
 
 
+
+
```
remove extra lines

carlescufi commented:
@XenuIsWatching please take another look
@anangl could you also review this please?

carlescufi commented:
It's taken a while but you got it merged @kurddt. Thanks for your patience.
